clients/spark/core/src/main/hadoop2/scala/io/treeverse/clients/conditional/S3ClientBuilder.scala

```diff
@@ -28,16 +30,49 @@ object S3ClientBuilder extends io.treeverse.clients.S3ClientBuilder {
         )
       } else None
 
+    initializeClient(configuration, credentialsProvider, region, bucket)
+  }
+
+  private def initializeClient(
+      config: ClientConfiguration,
+      credentialsProvider: Option[AWSCredentialsProvider],
+      region: String,
+      bucket: String
+  ): AmazonS3 = {
     val builder = AmazonS3ClientBuilder
       .standard()
-      .withClientConfiguration(configuration)
+      .withClientConfiguration(config)
       .withRegion(region)
-
     val builderWithCredentials = credentialsProvider match {
       case Some(cp) => builder.withCredentials(cp)
       case None     => builder
     }
+    val client = builderWithCredentials.build
+
+    if (validateClientCorrectness(client, bucket)) {
+      return client
+    }
+    val bucketRegion = getAWSS3Region(client, bucket)
+    builder.withRegion(bucketRegion).build
+  }
+
+  private def validateClientCorrectness(client: AmazonS3, bucket: String): Boolean = {
+    try {
+      client.headBucket(new HeadBucketRequest(bucket))
+      true
+    } catch {
+      case _: Exception => false
```

Comment:
This swallows all errors silently, which can make debugging harder.  Things that could make it a bit easier:

* Log the exception.
* Catch a more restricted set of exceptions, for example only AWS exceptions from `headBucket`.  Or report those differently.

Otherwise if I have the right region but my headBucket fails because of a sporadic issue and I have no permission to getBucketLocation, my GC fails (which is sort-of OK) but I don't understand why.